File: app/src/main/java/com/example/realweather/model/DisplayValueConverter.java
package com.example.realweather.model;

import android.util.ArrayMap;

import com.example.realweather.R;

public class DisplayValueConverter {

	private static final ArrayMap<String, Integer> ICON_MAPPING = createMapping();
	private boolean metricPreference;

	static ArrayMap<String, Integer> createMapping() {
		ArrayMap<String, Integer> mapping = new ArrayMap<>();
		mapping.put(WeatherDescription.CLEAR, R.drawable.ic_clear);
		mapping.put(WeatherDescription.LIGHT_RAIN, R.drawable.ic_light_rain);
		mapping.put(WeatherDescription.FOG, R.drawable.ic_fog);
		mapping.put(WeatherDescription.SNOW, R.drawable.ic_snow);
		mapping.put(WeatherDescription.STORM, R.drawable.ic_storm);
		mapping.put(WeatherDescription.FEW_CLOUD, R.drawable.ic_light_clouds);
		mapping.put(WeatherDescription.BROKEN_CLOUDS, R.drawable.ic_light_clouds);

		return mapping;
	}

	public DisplayValueConverter(boolean metricPreference) {
		this.metricPreference = metricPreference;
	}

	public int getIcon(String weatherDescription) {
		for (String key : ICON_MAPPING.keySet()) {
			if (key.contains(weatherDescription)) return ICON_MAPPING.get(key);
		}
		return R.drawable.example_appwidget_preview;
	}

	public String formatTemperature(double temperatureInKelvin) {
		return metricPreference ? kelvinToCelsius(temperatureInKelvin) : kelvinToFarenheit(temperatureInKelvin);
	}

	private String kelvinToCelsius(double temperatureInKelvin) {
		int result = (int) Math.round(temperatureInKelvin - 273.15);
		return result + "\u00B0";
	}

	private String kelvinToFarenheit(double temperatureInKelvin) {
		int result = (int) Math.round((temperatureInKelvin - 273.15) * 1.8 + 32);
		return result + "\u00B0";
	}

}

File: app/src/main/java/com/example/realweather/model/DisplayValueConverter.kt
package com.example.realweather.model

import android.util.ArrayMap
import com.example.realweather.R
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.format.TextStyle
import java.util.*
import kotlin.math.roundToInt

class DisplayValueConverter(private val metricPreference: Boolean) {
    fun getIcon(weatherDescription: String?): Int {
        for (key in ICON_MAPPING.keys) {
            weatherDescription?.let {
                if (weatherDescription.contains(key)) return ICON_MAPPING.getValue(key)
            }
        }
        return R.drawable.ic_light_clouds
    }

    fun formatTemperature(temperatureInKelvin: Double): String {
        return if (metricPreference) kelvinToCelsius(temperatureInKelvin) else kelvinToFahrenheit(
            temperatureInKelvin
        )
    }

    fun formatDescription(description: String): String {
        return description.split(' ').joinToString(" ") { it.replaceFirstChar(Char::titlecase) }
    }

    fun formatHumidity(humidity: Double): String {
        return "${humidity.roundToInt()}% RH"
    }


    private fun kelvinToCelsius(temperatureInKelvin: Double): String {
        val result = (temperatureInKelvin - 273.15).roundToInt()
        return result.toString() + "\u00B0"
    }

    private fun kelvinToFahrenheit(temperatureInKelvin: Double): String {
        val result = ((temperatureInKelvin - 273.15) * 1.8 + 32).roundToInt()
        return result.toString() + "\u00B0"
    }

    fun formatTime(time: String?): String {
        return LocalTime.parse(time).format(DateTimeFormatter.ofPattern("h:mma"))
    }

    fun formatDate(date: String?): String {
        val localDate = LocalDate.parse(
            date,
            DateTimeFormatter.ofPattern("yyyy-MM-dd")
        ) // LocalDate = 2010-02-23
        val weekday = localDate.dayOfWeek.getDisplayName(TextStyle.SHORT, Locale.US) // String = Tue
        val month = localDate.month.getDisplayName(TextStyle.SHORT, Locale.US)
        val day = localDate.dayOfMonth
        return "$weekday, $month $day"
    }

    companion object {
        private val ICON_MAPPING = createMapping().withDefault { R.drawable.ic_light_clouds }
        private fun createMapping(): ArrayMap<String, Int> {
            val mapping = ArrayMap<String, Int>()
            mapping[WeatherDescription.CLEAR] = R.drawable.ic_clear
            mapping[WeatherDescription.LIGHT_RAIN] = R.drawable.ic_light_rain
            mapping[WeatherDescription.FOG] = R.drawable.ic_fog
            mapping[WeatherDescription.SNOW] = R.drawable.ic_snow
            mapping[WeatherDescription.STORM] = R.drawable.ic_storm
            mapping[WeatherDescription.FEW_CLOUD] = R.drawable.ic_light_clouds
            mapping[WeatherDescription.BROKEN_CLOUDS] = R.drawable.ic_light_clouds
            return mapping
        }
    }
}
